### DISCUSSION TITLE
How to update model's dbName at runtime?

### DISCUSSION THREAD

**Discussion by nourgaserhd** [prisma/prisma#22532] [Q&A]
### Question

Hi! I have a bit of a complex database structure that I'm trying to integrate into. The database tables are prefixed and look something like this:
```
- userPrefix1_resource1
- userPrefix1_resource2
- userPrefix1_resource3
- userPrefix1_resourceN
- userPrefix2_resource1
- userPrefix2_resource2
- userPrefix2_resource3
- userPrefix2_resourceN
...rest_of_users
```
There's a different database to manage these prefixes, and each new user gets a record there, along with N new tables in this DB. I want to use Prisma as an ORM for this database in a new project (NestJS); I've scaffolded a set of N of these tables from this database and have what I call my "base schema" as attached to this thread. I want to create a middleware/extension to be able to use my generated Prisma client as such:
```
prisma.resourceX.someOperation();
```
I'll create a PrismaService to handle this, and inject into it another service responsible for getting the table prefix (userId) from the other database. The simplified flow will be like this:
```
1. NestJS server receives a request containing some userId.
2. Controller calls a service that needs to manage some resource, let's call it resourceX.
3. Said service calls PrismaService, along with the request context containing the userId.
4. PrismaService middleware calls the other database service to get the table prefix associated with this userId.
5. PrismaService middleware prepends the found prefix to the all table names in the query then executes it normally. 
```
**All I need help with is is step 5** (also 4 but it's NestJS-related; I don't know how I'll pass request.userId to the extension): the extension needs to, at runtime, modify the table name somehow. I tried this but it didn't work and _**the query executed as if I didn't do anything**_:
```typescript
const ExtdPrismaClient = class extends PrismaClient {
  constructor() {
    super();
    const client = new PrismaClient().$extends({
      query: {
        $allOperations({ model, args, query }) {
            const newName = "" //get the prefix by calling UserService.byId(req.userId).prefix;
            const oldName = client._runtimeDataModel.models[model].dbName; // this should be a loop and set the prefix for all models actually to handle joins
            client._runtimeDataModel.models[model].dbName = newName;
            // ^^^ didn't affect the query
            const result = query(args);
            client._runtimeDataModel.models[model].dbName = oldName;
            return result;
        },
      },
    });

    return client as PrismaClient;
  }
};
```
### Information about Prisma Schema, Client Queries and Environment (optional)

- schema
```prisma
generator client {
    provider = "prisma-client-js"
    output   = "../lib/client/tenants_data"
}

datasource db {
    provider = "mysql"
    url      = env("DB_URL")
}

model Resource1 {
    ID             Int     @id @default(autoincrement())
    Name           String  @db.VarChar(45)
    ...
}

model Resource2 {
    ID             Int     @id @default(autoincrement())
    Name           String  @db.VarChar(45)
    ...
}

model Resource3 {
    ID             Int     @id @default(autoincrement())
    Name           String  @db.VarChar(45)
    ...
}

...more resources similarly-defined
```
- Additional info
```
- OS: Pop!_OS 22.04
- Database: MySQL
- Node.js version: v21.2.0
```

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.7.1
@prisma/client          : Not found
Computed binaryTarget   : debian-openssl-1.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v21.2.0
Query Engine (Node-API) : libquery-engine 0ca5ccbcfa6bdc81c003cf549abe4269f59c41e5 (at .local/share/pnpm/global/5/.pnpm/@prisma+engines@5.7.1/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.0.x.so.node)
Schema Engine           : schema-engine-cli 0ca5ccbcfa6bdc81c003cf549abe4269f59c41e5 (at .local/share/pnpm/global/5/.pnpm/@prisma+engines@5.7.1/node_modules/@prisma/engines/schema-engine-debian-openssl-1.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.7.1-1.0ca5ccbcfa6bdc81c003cf549abe4269f59c41e5
Default Engines Hash    : 0ca5ccbcfa6bdc81c003cf549abe4269f59c41e5
Studio                  : 0.495.0
```

**Comment by nourgaserhd** [comment]
For reference, an old express server that uses Sequelize deals with it like this:
```js
// util function
function getPrefixedModel(tableName, tenantPrefix) {
  const model = allModels[tableName]; // imported
  model.tableName = tenantPrefix + "_" + tableName;
  model.sync({ force: false, logging: false });
  return model;
}
// model can be used like this 
let model = getPrefixedModel("resource1", "userPrefixFromReq");
await model.someOperation();
```
I guess it mutates the model, but that's okay since it'll always be called before using a model. But I really want to use Prisma for this project, so I hope someone can help me with a solution similar to this.

**Comment by ludralph** [comment]
Hi @nourgaserhd 👋 

Thank you for raising this question.

Currently, it is impossible to create or update a model name at runtime using Prisma ORM. We have an open [feature request](https://github.com/prisma/prisma/issues/21415) for creating models at runtime. We believe this is related to your use-case.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
